### DISCUSSION TITLE
Conditional `set` based on a parameter set on the host

### DISCUSSION THREAD

**Discussion by mikestreety** [deployphp/deployer#3373] [Help needed] [answered]
Hello,

(Using Deployer version 6)

I have several (20+) projects that all share similar environments and are set up to use one of two - either private servers or cPanels. We also have a mix of environments for individual projects (e.g. staging might be a VPS but live a cPanel). 

As all of our websites are set up the same, I wanted to share some config between them with a deployer meta package.

I was wondering - **how can I conditionally set certain parameters and fall back to defaults if not set?**

E.g., the composer path.

On the VPS, the composer path needs to be set to `/usr/local/bin/composer`, however on the cPanel environments I want it to work it out itself. I have set up the following in the hopes I can pick this up

```php
host('production')
	->set('ll_deployer_environment', 'cpanel') // can be cpanel or vps
	->set('deploy_path', '/dir')
;
```

I originally tried wrapping the `set` in an `if`, but it is never detected

```php
if (get('ll_deployer_environment') === 'vps') {
    set('bin/composer','/usr/local/bin/composer');
}
```

Next, I tried the `if` inside, which works except when the conditional is not set then there is an endless loop

```php
set('bin/composer', function () {
	if (get('ll_deployer_environment') === 'vps') {
		return '/usr/local/bin/composer';
	}

	return get('bin/composer');
});
```

**Lastly** i tried caching the value first, but that returns an error of `[Error] Object of class Closure could not be converted to string`

```php
$composer = get('bin/composer');
set('bin/composer', function () use ($composer) {
	if (get('ll_deployer_environment') === 'vps') {
		return '/usr/local/bin/composer';
	}

	return $composer;
});
```

I would like to be able to set something if a value is set on the host, and if not return the original function.

Please don't suggest:

- Trying to fix this specific composer issue (this is just an example, I have a few config settings)
- Upgrading Deployer (i'm using `deployer-extended` which is locked to 6 for now

Thanks for any suggestions, I will think of another way if this isn't possible.

Thanks

**Comment by antonmedv** [comment]
Use a separate task before deploy to set correct composer version.

**Reply by mikestreety** [reply]
Thanks! I just had the same thought about an hour ago. I'll post the solution once it's working but I will create a task called `prepare:environment` which can set the environment specific settings. 

Thanks again.

**Comment by mikestreety** [comment, marked as answer]
Thanks @antonmedv, worked perfectly.

For reference, this is what I ended up doing:

```php
/**
 * Set config based on environment
 */
task('environment:prepare', function () {
	switch (get('ll_deployer_environment')) {
		case 'vps':
			/**
			 * writable_use_sudo
			 * @package deployer
			 *
			 * Using sudo in writable commands?
			 */
			set('writable_use_sudo', true);

			/**
			* bin/composer
			* @package deployer
			*
			* Set default composer path
			*/
			set('bin/composer', '/usr/local/bin/composer');

			break;
		case 'cpanel':
			writeln('Environment: cPanel');

			/**
			 * writable_mode
			 * @package deployer
			 *
			 * What writeable mode should we use?
			 */
			set('writable_mode', 'chmod');

			break;

		default:
			writeln('Environment: Default');
			break;
	}
});

before('deploy:prepare', 'environment:prepare');
```

I can then do the following to use the specific config:

```php
host('production')
	->set('ll_deployer_environment', 'cpanel') // can be cpanel or vps
	->set('deploy_path', '/dir')
;
```
